### ISSUE TITLE
[Bug] Diffusion multi-update splits mini-batches contiguously; flow_grpo shuffles first

### ISSUE THREAD

**Issue by CjhHa1** [Tencent-Hunyuan/UniRL#323]
## Summary

Every diffusion recipe with `stack.num_updates_per_batch > 1` partitions the rollout into per-optimizer-step mini-batches by **contiguous index range**. The flow_grpo reference implementation shuffles the whole rollout *before* it splits. Because our samples arrive group-by-group, contiguous splitting hands each optimizer step a set of **complete** prompt groups, while flow_grpo hands each step a random subset that cuts groups apart.

Neither is obviously wrong, but the published flow_grpo curves were produced with the shuffle and we deviate from it silently.

Scope: **72** recipes under `examples/diffusion/`, `examples/unified_model/`, `examples/pe/` run at `num_updates_per_batch >= 2`. The 16 `examples/ar/*` recipes at `>= 2` are **not** affected (see "AR is already at parity").

## What flow_grpo does

`scripts/train_sd3.py` (yifan123/flow_grpo), inside the outer epoch:

```python
for inner_epoch in range(config.train.num_inner_epochs):
    # shuffle samples along batch dimension
    perm = torch.randperm(total_batch_size, device=accelerator.device)
    samples = {k: v[perm] for k, v in samples.items()}

    # rebatch for training
    samples_batched = {
        k: v.reshape(-1, total_batch_size//config.sample.num_batches_per_epoch, *v.shape[1:])
        for k, v in samples.items()
    }
```

Optimizer steps per rollout is `samples_per_epoch // total_train_batch_size` times `num_inner_epochs`. With the `config/base.py` defaults (`sample.num_batches_per_epoch = 2`, `train.batch_size = 1`, `gradient_accumulation_steps = 1`, `num_inner_epochs = 1`) that is 2 — the same 2 our diffusion recipes carry, and it comes from `num_batches_per_epoch`, not from re-epoching. The `randperm` runs before the reshape, so it is what decides *which* samples land in step 1 vs step 2.

## What we do

`unirl/train/stack/planner/types.py`:

```python
def _update_ranges(*, total_size: int, num_updates: int) -> Tuple[Range, ...]:
    ...
    update_size = total // n
    return tuple((i * update_size, (i + 1) * update_size) for i in range(n))
```

Contiguous, no permutation. Both planners route through it: `count.py` directly, and `packed.py` slices `samples[u_start:u_end]` first and only reorders *within* each update for token packing. So update membership is contiguous in original order either way.

## Why the membership matters

Samples reach the stack group-contiguous (`Sample.concat` over per-prompt groups). With 4 prompts x 4 samples and `num_updates_per_batch=2`:

```
index : 0  1  2  3 | 4  5  6  7 | 8  9 10 11 | 12 13 14 15
prompt: A  A  A  A | B  B  B  B | C  C  C  C |  D  D  D  D

ours      : update 1 = A,B complete        update 2 = C,D complete
flow_grpo : update 1 = random 8 of 16      update 2 = the other 8
            (partial groups, all 4 prompts in both)
```

GRPO advantages are group-normalized *before* the split, so a complete group's advantages sum to ~0. Ours therefore gives every optimizer step a balanced within-group contrast, at the cost of each step seeing a disjoint prompt subset. flow_grpo gives every step all prompts but partial groups, so one step can be net-positive or net-negative (e.g. a group's two best samples both land in step 1). Across the N steps that averages out; within a step it does not.

Prompt order entering a rollout is already random on both sides (`run.shuffle` defaults to `True` in `MultimodalRLDataSource`), so at *prompt* granularity the two schemes are statistically equivalent. The remaining difference is group integrity.

## AR is already at parity

verl exposes this as an explicit knob and defaults it off:

```python
# verl/workers/config/actor.py
ppo_epochs: int = 1
shuffle: bool = False
```

Our contiguous split matches verl's default, so the AR recipes are at reference parity as-is. (`data.shuffle: True` in verl is the dataloader-level prompt shuffle — a different layer, which we also do.) This issue is diffusion-only.

## What to decide

1. **Match flow_grpo** — permute the track before `_update_ranges`, seeded so runs stay reproducible. Restores parity for all 72 recipes; changes numerics on every existing diffusion run.
2. **Make it a knob** — a `TrainStack(shuffle_updates=...)` mirroring verl's `actor.shuffle`, default `False` to preserve today's behavior, set `True` only in the recipes that claim flow_grpo parity.
3. **Keep contiguous and document it** — argue that whole-group mini-batches are better conditioned for group-normalized advantages, and record the deviation in `unirl/train/readme.md`.

Option 2 looks best to me: it is the only one that both preserves current numerics and makes a parity run requestable. But it needs an SD3 A/B first to establish whether the difference is measurable at all — if it is not, option 3 with a documented note is cheaper.

## Verified (was: not verified)

Resolved in the comment below. Summary: **DanceGRPO and MixGRPO do not shuffle the batch axis** — they shuffle the *time* axis (per-sample `randperm(num_timesteps)`), so our contiguous split is already at parity with them and their recipes must be excluded from any batch-axis fix. **FlowDPPO** has no external reference implementation and inherits whatever is decided here. This narrows the batch-axis gap to the flow_grpo-parity recipes only, and makes option 2 the clear choice over option 1.

The comment also records a **separate time-axis gap**: our `AllSDEScheduler` draws the trained SDE steps once per rollout and shares them across the whole batch, while DanceGRPO/MixGRPO draw per sample. That needs its own issue.

**Comment by CjhHa1** [comment]
Checked the three unverified references. The picture is more nuanced than the issue body assumes: **there are two independent shuffle axes, and the upstreams disagree about which one they randomize.**

| | batch axis (which samples enter which optimizer step) | time axis (which SDE steps of a sample get trained) |
|---|---|---|
| flow_grpo | **shuffles** — `randperm(total_batch_size)` before the reshape | no randomization; deterministic prefix via `train.timestep_fraction` |
| DanceGRPO | no shuffle — `v.reshape(-1, config.train.batch_size, ...)` straight through | **shuffles, per sample independently** — `randperm(num_timesteps)` for each of `total_batch_size` |
| MixGRPO | no shuffle — processes one sample per step (`v.unsqueeze(1)`) | **shuffles, per sample independently**, only when `training_strategy == "all"` |
| ours | no shuffle — `_update_ranges` contiguous | randomized, but **one draw shared by the whole batch**, seeded by `rollout_id` |

### DanceGRPO (`fastvideo/train_grpo_sd.py`)

```python
for inner_epoch in range(config.train.num_inner_epochs):
    # shuffle along time dimension independently for each sample
    perms = torch.stack(
        [
            torch.randperm(num_timesteps, device=accelerator.device)
            for _ in range(total_batch_size)
        ]
    )
    for key in ["timesteps", "latents", "next_latents", "log_probs"]:
        samples[key] = samples[key][
            torch.arange(total_batch_size, device=accelerator.device)[:, None],
            perms,
        ]

    # rebatch for training
    samples_batched = {
        k: v.reshape(-1, config.train.batch_size, *v.shape[1:])
        for k, v in samples.items()
    }
```

The comment says it outright: *shuffle along time dimension independently for each sample*. There is no batch-dimension permutation anywhere in `train_grpo_sd.py` or `train_grpo_flux.py` — the only `randperm` in either file is over `num_timesteps`. Otherwise the loop is the same DDPO lineage as flow_grpo (`assert samples_per_epoch % total_train_batch_size == 0`, `num_inner_epochs`).

### MixGRPO (`fastvideo/train_grpo_flux.py`)

```python
if args.training_strategy == "all":
    perms = torch.stack(
        [
            torch.randperm(len(samples["timesteps"][0]))
            for _ in range(batch_size)
        ]
    ).to(device)
    for key in ["timesteps", "latents", "next_latents", "log_probs"]:
        samples[key] = samples[key][
            torch.arange(batch_size).to(device)[:, None],
            perms,
        ]
```

Same per-sample time-axis shuffle as DanceGRPO, gated on the non-window strategy. No batch-axis permutation.

### FlowDPPO

No external reference implementation to be at parity with. `FlowDPPO/README.md` points the loss at `unirl/algorithms/flowdppo.py`, cites the paper without a code release, and lists the official `Tencent-Hunyuan-Multimodal-RL` adapters — this is original work here, and the README states it reuses FlowGRPO's SDE rollout. It inherits whatever this repo decides; it imposes no independent constraint.

## What this changes about this issue

**Scope shrinks on the batch axis.** DanceGRPO and MixGRPO do *not* shuffle the batch dimension, so our contiguous `_update_ranges` is already at parity with them. Of the 72 recipes, the batch-axis gap only applies to the ones claiming flow_grpo parity — the `sd3_*`, `qwen_image_*`, `z_image_*`, `bagel_*`, `flux2_klein_*`, `hunyuan_video*`, `wan*` non-dance/non-mix set. The `*_dancegrpo*` and `*_mixgrpo*` recipes should be excluded from any batch-axis fix, which makes option 2 (a per-recipe `shuffle_updates` knob) clearly better than option 1 (change it globally).

**A second, separate gap appears on the time axis.** Our `AllSDEScheduler.get_sde_indices` draws once per rollout and shares that set across the entire batch:

```python
seed = 0 if step is None else int(step)
rng = np.random.default_rng(seed)
chosen = rng.choice(pool, size=self.num_sde_steps, replace=False)
```

called as `sde_indices = diffusion.resolve_sde_indices(rollout_id)`. So every sample in a rollout trains on the *same* `num_sde_steps` SDE steps. DanceGRPO and MixGRPO draw independently per sample, so their batch collectively covers far more of the denoising trajectory per update, while ours concentrates the gradient on one shared subset (different subset each rollout, so it averages over training, but not within an update).

That is a distinct question from mini-batch membership — different axis, different upstreams, different fix (per-sample index draws would have to be threaded through the segment, not just the planner) — so it should get its own issue rather than being folded in here.
